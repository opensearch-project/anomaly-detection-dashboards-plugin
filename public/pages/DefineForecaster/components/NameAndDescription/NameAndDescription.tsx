/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { EuiCompressedFieldText, EuiCompressedTextArea, EuiSpacer, EuiText } from '@elastic/eui';
import { Field, FieldProps } from 'formik';
import React, { Fragment } from 'react';
import { getError, isInvalid } from '../../../../utils/utils';
import { validateForecasterDesc } from './utils/validation';
import { FormattedFormRow } from '../../../../components/FormattedFormRow/FormattedFormRow';

interface NameAndDescriptionProps {
  onValidateForecasterName: (forecasterName: string) => Promise<any>;
  omitTitle?: boolean;
  isEditable?: boolean;
}

function NameAndDescription({ onValidateForecasterName, omitTitle = false, isEditable = true }: NameAndDescriptionProps) {
  console.log('isEditable', isEditable);
  return (
    <Fragment>
      {!omitTitle && (
        <>
          <EuiText>
            <h4>Details</h4>
          </EuiText>
          <EuiSpacer size="m" />
        </>
      )}
      {/* FIXME: Add hint text to emphasize that forecaster name must be unique across the system */}
      <Field 
        name="name" 
        validate={isEditable ? onValidateForecasterName : undefined}  // Only validate if editable
      >
        {({ field, form }: FieldProps) => (
          <FormattedFormRow
            title="Name"
            isInvalid={isInvalid(field.name, form)}
            error={getError(field.name, form)}
            helpText={`Specify a unique name. Must contain 1-64 characters. Valid characters are
                a-z, A-Z, 0-9, -(hyphen), _(underscore) and .(period).`}
          >
            {/* FIXME: even though readOnly is more desirable, but to be consistent with other components 
                who don't have readOnly mode, keep disabled instead */}
            <EuiCompressedFieldText
              data-test-subj="forecasterNameTextInput"
              name="forecasterName"
              id="forecasterName"
              placeholder="Enter forecaster name"
              isInvalid={isInvalid(field.name, form)}
              disabled={!isEditable}
              {...field}
              // Ensure value is always a defined string to prevent React warning about
              // switching between controlled/uncontrolled input. Empty string fallback
              // is required because undefined/null values can cause this switch.
              value={field.value || ''}
            />
          </FormattedFormRow>
        )}
      </Field>

      <Field 
        name="description" 
        validate={isEditable ? validateForecasterDesc : undefined}  // Only validate if editable
      >
        {({ field, form }: FieldProps) => (
          <FormattedFormRow
            formattedTitle={
              <p>
                Description <span className="optional">- optional</span>
              </p>
            }
            isInvalid={isInvalid(field.name, form)}
            error={getError(field.name, form)}
          >
            {/* FIXME: even though readOnly is more desirable, but to be consistent with other components 
                who don't have readOnly mode, keep disabled instead */}
            <EuiCompressedTextArea
              data-test-subj="forecasterDescriptionTextInput"
              name="forecasterDescription"
              id="forecasterDescription"
              rows={3}
              placeholder="Describe the forecaster"
              disabled={!isEditable}
              {...field}
              isInvalid={isInvalid(field.name, form)}
            />
          </FormattedFormRow>
        )}
      </Field>
    </Fragment>
  );
}

export default NameAndDescription;
