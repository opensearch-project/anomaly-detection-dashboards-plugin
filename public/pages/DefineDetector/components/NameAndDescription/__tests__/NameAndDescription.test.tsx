/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import NameAndDescription from '../NameAndDescription';

import { Formik } from 'formik';

describe('<NameAndDescription /> spec', () => {
  test('renders the component', () => {
    const { container } = render(
      <Formik initialValues={{ detectorName: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <NameAndDescription onValidateDetectorName={jest.fn()} />
          </div>
        )}
      </Formik>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  test('shows error for detector name input when toggling focus/blur', async () => {
    const handleValidateName = jest.fn().mockImplementation(() => {
      throw 'Required';
    });
    const { queryByText, findByText, getByPlaceholderText } = render(
      <Formik initialValues={{ name: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <NameAndDescription onValidateDetectorName={handleValidateName} />
          </div>
        )}
      </Formik>
    );
    expect(queryByText('Required')).toBeNull();
    fireEvent.focus(getByPlaceholderText('Enter detector name'));

    fireEvent.blur(getByPlaceholderText('Enter detector name'));
    expect(handleValidateName).toHaveBeenCalledWith('');
    expect(handleValidateName).toHaveBeenCalledTimes(1);
    expect(findByText('Required')).not.toBeNull();
  });
  test('shows error for detector description input when toggling focus/bur', async () => {
    const { queryByText, findByText, getByPlaceholderText } = render(
      <Formik
        initialValues={{ name: '', description: '' }}
        onSubmit={jest.fn()}
      >
        {() => (
          <div>
            <NameAndDescription onValidateDetectorName={jest.fn()} />
          </div>
        )}
      </Formik>
    );
    expect(queryByText('Required')).toBeNull();
    fireEvent.focus(getByPlaceholderText('Describe the detector'));
    fireEvent.blur(getByPlaceholderText('Describe the detector'));
    expect(findByText('Required')).not.toBeNull();
  });
});
