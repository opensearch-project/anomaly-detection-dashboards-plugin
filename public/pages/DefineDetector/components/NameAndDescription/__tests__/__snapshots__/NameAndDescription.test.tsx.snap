// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NameAndDescription /> spec renders the component 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium"
    style="padding: 20px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Detector details
        </h3>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem content-panel-subTitle"
            style="line-height: normal; max-width: 75%;"
          />
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
    </div>
    <div>
      <hr
        class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
      />
      <div
        style="padding: 10px 0px;"
      >
        <div
          class="euiFormRow"
          hint="Specify a unique and descriptive name that is easy to recognize."
          id="random_html_id-row"
          title="Name"
        >
          <div
            class="euiFormRow__labelWrapper"
          >
            <label
              aria-invalid="false"
              class="euiFormLabel euiFormRow__label"
              for="random_html_id"
            >
              <div
                style="line-height: 8px;"
              >
                <p>
                  Name
                </p>
                <br />
                <div
                  class="euiText euiText--medium sublabel"
                  style="max-width: 400px;"
                >
                  Specify a unique and descriptive name that is easy to recognize.
                </div>
              </div>
            </label>
          </div>
          <div
            class="euiFormRow__fieldWrapper"
          >
            <div
              class="euiFormControlLayout"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <input
                  aria-describedby="random_html_id-help"
                  class="euiFieldText"
                  id="random_html_id"
                  name="name"
                  placeholder="Enter detector name"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="euiFormHelpText euiFormRow__text"
              id="random_html_id-help"
            >
              Detector name must contain 1-64 characters. Valid characters are
                a-z, A-Z, 0-9, -(hyphen) and _(underscore).
            </div>
          </div>
        </div>
        <div
          class="euiFormRow"
          hint="Describe the purpose of the detector."
          id="random_html_id-row"
        >
          <div
            class="euiFormRow__labelWrapper"
          >
            <label
              aria-invalid="false"
              class="euiFormLabel euiFormRow__label"
              for="random_html_id"
            >
              <div
                style="line-height: 8px;"
              >
                <p>
                  Description 
                  <span
                    class="optional"
                  >
                    - optional
                  </span>
                </p>
                <br />
                <div
                  class="euiText euiText--medium sublabel"
                  style="max-width: 400px;"
                >
                  Describe the purpose of the detector.
                </div>
              </div>
            </label>
          </div>
          <div
            class="euiFormRow__fieldWrapper"
          >
            <textarea
              class="euiTextArea euiTextArea--resizeVertical"
              id="random_html_id"
              name="description"
              placeholder="Describe the detector"
              rows="3"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
